keps/sig-node/20190528-node-maintenance-controller.md

```diff
@@ -0,0 +1,190 @@
+- The controller watches and acts upon maintenance CRs to put/remove a node on/from maintenance.
+- Status report of the drain (embodied in the maintenance CR).
+- Custom taints and labels added to the node going under maintenance.
+- Configurable time-out(time buffer) for the drain process.
+
+### Non-Goals
+- "Dry run" mode to detect what is the best node / N nodes to drain
+
+## Proposal
+
+Node maintenance controller watches for new or deleted custom resources e.g. `NodeMaintenance` which indicate that a node in the cluster should either:
+ - `NodeMaintenance` CR created: move node into maintenance, cordon the node - set it as unscheduleble and evict the pods (which can be evicted) from that node.
+  - `NodeMaintenance` CR deleted: remove pod from maintenance and uncordon the node - set it as scheduleble.
+
+Current implementation and design can be found under Kubevirt:
+https://github.com/kubevirt/node-maintenance-operator 
+
+### User Stories
+
+#### Story 1
+As a third party developer/operator I want to evacuate my workloads appropriately when a node goes into maintenance mode, but I do not want to evacuate them when the node is cordoned.
+
+- Today we can not differentiate between cordon and drain using cluster information
+
+- Adding taints/labels as part of the evacuation process can help performing additional work before a pod is removed from a node.
+
+#### Story 2
+As a third party developer/operator I would like to activate the node back simply by removing the maintenance mode CR.
+
+
+#### Story 3
+As an operator I need a reliable way to request to put a node into maintenance mode and understand if it is - or not - in order to perform i.e. hardware maintenance work.
+
+- CR needs to report pending/running/success/failure as well as informative conditions in the CR’s Status section. The conditions should provide information about what is blocking maintenance node in the event maintenance mode is can not proceed.
+
+#### Story 4
+As an operator I need to put a number of nodes i.e. a rack of nodes into node maintenance.
+
+- As a cluster admin i would like to shutdown a rack (due to overheating problem) to do that I would need to move hosts to maintenance.
+
+#### Story 5
+- As a cluster admin, i would like to perform a firmware update on the nodes.
+
+- As a cluster admin i would like to perform a bios update for the nodes in the cluster.
+
+#### Story 6
+- As a developer/operator I would like a way to invoke node maintenance (drain) not only from CLI but also from UI.
+
+### Implementation Details/Notes/Constraints
+
+- Custom taints added to the node by the controller might be overun by other controllers such as the node life cycle controller for the `TaintNodesByCondition` feature or the node controller for the `TaintBasedEvictions` feature.
```

Comment:
By "overrun" do you mean "overwritten"? If you are using a custom taint key, I would not expect other controllers to modify your taint.